server/plugin.ts

```diff
@@ -87,107 +83,6 @@ export class SecurityPlugin implements Plugin<SecurityPluginSetup, SecurityPlugi
-      return pathsToIgnore.includes(url.pathname?.split('/').pop() || '');
-    } catch (error: any) {
-      this.logger.error(`Could not parse the referer for the capabilites: ${error.stack}`);
-    }
-  }
-
-  isReadOnlyTenant(authInfo: any): boolean {
-    const currentTenant = authInfo.user_requested_tenant || globalTenantName;
-
-    // private tenant is not affected
-    if (isPrivateTenant(currentTenant)) {
-      return false;
-    }
-
-    return authInfo.tenants[currentTenant] !== true;
-  }
-
-  toggleReadOnlyCapabilities(capabilities: any): Partial<Capabilities> {
-    return Object.entries(capabilities).reduce((acc, cur) => {
-      const [key, value] = cur;
-
-      return { ...acc, [key]: capabilities.hide_for_read_only.includes(key) ? false : value };
-    }, {});
-  }
-
-  toggleForReadOnlyTenant(uiCapabilities: Capabilities): Partial<Capabilities> {
-    const defaultTenantOnlyCapabilities = Object.entries(uiCapabilities).reduce((acc, cur) => {
-      const [key, value] = cur;
-
-      if (!value.hide_for_read_only) {
-        return acc;
-      }
-
-      const updatedValue = this.toggleReadOnlyCapabilities(value);
-
-      return { ...acc, [key]: updatedValue };
-    }, {});
-
-    const finalCapabilities = merge(uiCapabilities, defaultTenantOnlyCapabilities);
-
-    return finalCapabilities;
-  }
-
-  capabilitiesSwitcher(
-    securityClient: SecurityClient,
-    auth: IAuthenticationType,
-    securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>
-  ): CapabilitiesSwitcher {
-    return async (
-      request: OpenSearchDashboardsRequest,
-      uiCapabilities: Capabilities
-    ): Promise<Partial<Capabilities>> => {
-      // omit for anonymous pages to avoid authentication errors
-      if (this.isAnonymousPage(request)) {
-        return uiCapabilities;
-      }
-
-      try {
-        const cookie = await securitySessionStorageFactory.asScoped(request).get();
-        let headers = request.headers;
-
-        if (!auth.requestIncludesAuthInfo(request) && cookie) {
-          headers = auth.buildAuthHeaderFromCookie(cookie, request);
-        }
-
-        const authInfo = await securityClient.authinfo(request, headers);
-
-        if (!authInfo.user_requested_tenant && cookie) {
-          authInfo.user_requested_tenant = cookie.tenant;
-        }
-
-        if (this.isReadOnlyTenant(authInfo)) {
-          return this.toggleForReadOnlyTenant(uiCapabilities);
-        }
-      } catch (error: any) {
-        this.logger.error(`Could not check auth info: ${error.stack}`);
-      }
-
-      return uiCapabilities;
-    };
-  }
-
-  registerSwitcher(
-    core: CoreSetup,
-    securityClient: SecurityClient,
-    auth: IAuthenticationType,
-    securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>
-  ) {
-    core.capabilities.registerSwitcher(
-      this.capabilitiesSwitcher(securityClient, auth, securitySessionStorageFactory)
-    );
-  }
```

Comment:
For my understanding, what is the difference between registering a switcher vs the new way?

Reply:
@DarshitChanpura the `registerSwitcher` way introduced a `hide_for_read_only` which was extending capabilities type and was a messy way of saying "change it for me". With a new approach, we can use the existing `registerSwitcher` and separate logic to check if the current tenant is read-only without bruising capabilities definition. We're then even more able to define behaviors of each plugin in case of read-only.